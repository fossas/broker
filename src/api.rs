//! Interaction with remote APIs live here.
//! See each submodule for details.

pub mod code_host;
pub mod fossa;
pub mod http;
pub mod ssh;
