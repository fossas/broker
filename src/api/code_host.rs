//! This module provides functionality for integrating with external code hosts.

use std::time::Duration;

use derive_more::{AsRef, Display, From};
use derive_new::new;
use error_stack::{IntoReport, Report, ResultExt};
use getset::{CopyGetters, Getters};
use url::Url;

use crate::ext::error_stack::{DescribeContext, ErrorHelper};

pub mod git;

/// Errors that are possibly surfaced during validation of config values.
#[derive(Debug, thiserror::Error)]
pub enum ValidationError {
    /// Poll interval is parsed from a user-provided string.
    #[error("validate poll interval")]
    PollInterval,

    /// The provided URL is not valid.
    #[error("validate endpoint URL")]
    ValidateEndpoint,
}

/// Validated config values for external code host integrations.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct Config(Vec<Integration>);

/// Validated endpoint for a code host.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, From, Display, new)]
pub struct Endpoint(Url);

impl TryFrom<String> for Endpoint {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        Url::parse(&input)
            .into_report()
            .describe_lazy(|| format!("provided input: '{input}'"))
            .help("the url provided must be absolute and must contain the protocol")
            .change_context(ValidationError::ValidateEndpoint)
            .map(Endpoint)
    }
}

/// Each integration has metadata (for example, how often it should be polled)
/// along with its protocol (which describes how to download the code so it can be analyzed).
///
/// This type stores this combination of data.
#[derive(Debug, Clone, PartialEq, Eq, Getters, CopyGetters, new)]
pub struct Integration {
    /// The interval at which Broker should poll the remote code host for whether the code has changed.
    #[getset(get_copy = "pub")]
    poll_interval: PollInterval,

    /// The protocol Broker uses to communicate with the remote code host.
    #[getset(get = "pub")]
    protocol: Protocol,
}

/// Code is stored in many kinds of locations, from git repos to
/// random FTP sites to DevOps hosts like GitHub.
///
/// To handle this variety, Broker uses a predefined list
/// of supported protocols (this type),
/// which are specialized with configuration unique to those integrations.
#[derive(Debug, Clone, PartialEq, Eq, From, new)]
pub enum Protocol {
    /// Integration with a code host using the git protocol.
    Git(git::Transport),
}

/// Specifies the maximum age for an observability artifact.
#[derive(Debug, Copy, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct PollInterval(Duration);

impl TryFrom<String> for PollInterval {
    type Error = Report<ValidationError>;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        parse_duration::parse(&value)
            .into_report()
            .change_context(ValidationError::PollInterval)
            .describe_lazy(|| format!("provided value: {value}"))
            .map(PollInterval)
    }
}
