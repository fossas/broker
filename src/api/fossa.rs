//! Interactions and data types for the FOSSA API live here.

use derive_more::{AsRef, Display, From};
use derive_new::new;
use error_stack::{IntoReport, Report, ResultExt};
use getset::Getters;
use url::Url;

use crate::ext::{
    error_stack::{DescribeContext, ErrorHelper},
    secrecy::ComparableSecretString,
};

/// Errors that are possibly surfaced during validation of config values.
#[derive(Debug, thiserror::Error)]
pub enum ValidationError {
    /// The provided URL is not valid.
    #[error("validate endpoint URL")]
    Endpoint,

    /// The provided API key is not valid.
    #[error("validate API key")]
    ApiKey,

    /// The value provided to parse is empty.
    #[error("provided value is empty")]
    ValueEmpty,
}

/// Validated config values for the FOSSA API.
#[derive(Debug, Clone, PartialEq, Eq, Getters, new)]
#[getset(get = "pub")]
pub struct Config {
    /// The endpoint for the FOSSA backend.
    endpoint: Endpoint,

    /// The key used when interacting with the FOSSA backend.
    key: Key,
}

/// The URL to the FOSSA endpoint.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, Display, From, new)]
pub struct Endpoint(Url);

impl TryFrom<String> for Endpoint {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        Url::parse(&input)
            .into_report()
            .describe_lazy(|| format!("provided input: '{input}'"))
            .help("the url provided must be absolute and must contain the protocol, for example 'https://app.fossa.com'")
            .change_context(ValidationError::Endpoint)
            .map(Endpoint)
    }
}

/// The FOSSA API key.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct Key(ComparableSecretString);

impl TryFrom<String> for Key {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        if input.is_empty() {
            Err(Report::new(ValidationError::ValueEmpty))
                .describe_lazy(|| format!("provided input: '{input}'"))
                .help("use an API key from FOSSA here: https://app.fossa.com/account/settings/integrations/api_tokens")
                .change_context(ValidationError::ApiKey)
        } else {
            let secret = ComparableSecretString::from(input);
            Ok(Key(secret))
        }
    }
}
