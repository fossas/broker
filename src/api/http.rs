//! Interact with remote services over HTTP!

use std::fmt::Display;

use derive_more::From;
use derive_new::new;
use serde::{Deserialize, Serialize};

use crate::ext::secrecy::ComparableSecretString;

/// HTTP authentication can be performed either with a header or via 'HTTP Basic'.
#[derive(Debug, Clone, PartialEq, Eq, From, Deserialize, Serialize, new)]
pub enum Auth {
    /// Uses a header value for authentication.
    Header(ComparableSecretString),

    /// Uses HTTP Basic to perform authentication.
    Basic {
        /// The username for authentication.
        username: String,

        /// The password for authentication.
        password: ComparableSecretString,
    },
}

impl Display for Auth {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Auth::Header(header) => write!(f, "authorization header {header}"),
            Auth::Basic { username, password } => {
                write!(f, "username '{username}' and password '{password}'")
            }
        }
    }
}
