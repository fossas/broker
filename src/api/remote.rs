//! This module provides functionality for integrating with "remotes", which are external code hosts.
//!
//! # "Remote" as a concept
//!
//! An example of a [`Remote`] is Github; it's a place where code is stored remotely.
//! [`Remote`] as a concept is abstracted over protocol implementation:
//! for example, a code host accessed via `ssh` or via `http` are both simply a [`Remote`].
//!
//! Their connection protocol (and any specifics, like authentication) is then specified via
//! [`Protocol`], which is usually wrapped inside an [`Integration`], forming the primary interaction
//! point for this module.

use std::time::Duration;

use derive_more::{AsRef, Display, From};
use derive_new::new;
use error_stack::{report, IntoReport, Report, ResultExt};
use getset::{CopyGetters, Getters};
use humantime::parse_duration;
use url;

use crate::ext::error_stack::{DescribeContext, ErrorHelper};

pub mod git;

/// Errors that are possibly surfaced during validation of config values.
#[derive(Debug, thiserror::Error)]
pub enum ValidationError {
    /// Poll interval is parsed from a user-provided string.
    #[error("validate poll interval")]
    PollInterval,

    /// The provided remote is not valid.
    #[error("validate remote location")]
    Remote,

    /// The provided value is empty.
    #[error("value is empty")]
    ValueEmpty,
}

/// Validated config values for external code host integrations.
#[derive(Debug, Default, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct Config(Vec<Integration>);

/// Validated remote location for a code host.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, Display, new)]
pub struct Remote(String);

impl Remote {
    /// parses a Remote and returns a Result<Url>
    pub fn parse(&self) -> Result<url::Url, url::ParseError> {
        url::Url::parse(&self.0)
    }
}

impl TryFrom<String> for Remote {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        // Different remotes have different semantics-
        // we can't guarantee this is actually a well formatted URL.
        // Just validate that it's not empty.
        if input.is_empty() {
            Err(report!(ValidationError::ValueEmpty))
                .describe_lazy(|| format!("provided input: '{input}'"))
        } else {
            Ok(Remote(input))
        }
        .help("the remote location may not be empty")
        .change_context(ValidationError::Remote)
    }
}

/// Each integration has metadata (for example, how often it should be polled)
/// along with its protocol (which describes how to download the code so it can be analyzed).
///
/// This type stores this combination of data.
#[derive(Debug, Clone, PartialEq, Eq, Getters, CopyGetters, new)]
pub struct Integration {
    /// The interval at which Broker should poll the remote code host for whether the code has changed.
    #[getset(get_copy = "pub")]
    poll_interval: PollInterval,

    /// The protocol Broker uses to communicate with the remote code host.
    #[getset(get = "pub")]
    protocol: Protocol,
}

/// Code is stored in many kinds of locations, from git repos to
/// random FTP sites to DevOps hosts like GitHub.
///
/// To handle this variety, Broker uses a predefined list
/// of supported protocols (this type),
/// which are specialized with configuration unique to those integrations.
#[derive(Debug, Clone, PartialEq, Eq, From, new)]
pub enum Protocol {
    /// Integration with a code host using the git protocol.
    Git(git::Transport),
}

/// Specifies the maximum age for an observability artifact.
#[derive(Debug, Copy, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct PollInterval(Duration);

impl TryFrom<String> for PollInterval {
    type Error = Report<ValidationError>;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        parse_duration(&value)
            .into_report()
            .change_context(ValidationError::PollInterval)
            .describe_lazy(|| format!("provided value: {value}"))
            .map(PollInterval)
    }
}
