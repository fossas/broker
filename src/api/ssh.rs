//! Interact with remote services over SSH.

use std::{fmt::Display, path::PathBuf};

use derive_more::From;
use derive_new::new;
use serde::{Deserialize, Serialize};

use crate::ext::secrecy::ComparableSecretString;

/// SSH authentication can be performed either with a key file or with a static key.
#[derive(Debug, Clone, PartialEq, Eq, From, Deserialize, Serialize, new)]
pub enum Auth {
    /// Uses a key file on disk to perform authentication.
    KeyFile(PathBuf),

    /// Uses a value stored in the config file (and therefore in memory) to perform authentication.
    KeyValue(ComparableSecretString),
}

impl Display for Auth {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Auth::KeyFile(path) => write!(f, "key file at '{}'", path.display()),
            Auth::KeyValue(key) => write!(f, "private key {key}"),
        }
    }
}
