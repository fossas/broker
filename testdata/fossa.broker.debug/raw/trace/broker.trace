{"timestamp":"2023-04-05T16:30:24.086526Z","level":"INFO","fields":{"message":"Debug artifacts being stored in '/Users/me/.config/fossa/broker/debug/trace'"},"target":"broker::debug"}
