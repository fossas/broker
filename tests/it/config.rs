use broker::api::{self, remote};

use crate::{assert_error_stack_snapshot, helper::gen, load_config, load_config_err};

#[tokio::test]
async fn test_fossa_api_values() {
    let (_, conf) = load_config!().await;

    assert_eq!(conf.fossa_api().key(), &gen::fossa_api_key("abcd1234"),);
    assert_eq!(
        conf.fossa_api().endpoint(),
        &gen::fossa_api_endpoint("https://app.fossa.com"),
    );
}

#[tokio::test]
async fn test_debug_values() {
    let (_, conf) = load_config!().await;

    assert_eq!(
        conf.debug().location(),
        &gen::debug_root("/home/me/.config/fossa/broker/debugging/"),
    );
    assert_eq!(
        conf.debug().retention().days(),
        gen::debug_artifact_retention_count(3),
    );
}

#[tokio::test]
async fn test_debug_values_default_retention() {
    let (_, conf) = load_config!(
        "testdata/config/basic-retention-default.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    assert_eq!(
        conf.debug().location(),
        &gen::debug_root("/home/me/.fossa/broker/debugging/"),
    );
    assert_eq!(
        conf.debug().retention().days(),
        gen::debug_artifact_retention_count(7),
    );
}

#[tokio::test]
async fn test_debug_values_retention_malformed() {
    let (config_path, err) = load_config_err!(
        "testdata/config/basic-retention-malformed.yml",
        "testdata/database/empty.sqlite"
    )
    .await;
    assert_error_stack_snapshot!(&config_path, err);
}

#[tokio::test]
async fn test_debug_values_retention_invalid() {
    let (config_file_path, err) = load_config_err!(
        "testdata/config/basic-retention-invalid.yml",
        "testdata/database/empty.sqlite"
    )
    .await;
    assert_error_stack_snapshot!(&config_file_path, err);
}

#[tokio::test]
async fn test_one_integration() {
    let (_, conf) = load_config!().await;

    let mut integrations = conf.integrations().as_ref().iter();
    let Some(_) = integrations.next() else { panic!("must have parsed at least one integration") };
    let None = integrations.next() else { panic!("must have parsed exactly one integration") };
}

#[tokio::test]
async fn test_integration_git_ssh_key_file() {
    let (_, conf) = load_config!().await;

    let Some(integration) = conf.integrations().as_ref().iter().next() else { panic!("must have parsed at least one integration") };
    assert_eq!(integration.poll_interval(), gen::code_poll_interval("1h"));

    let remote::Protocol::Git(remote::git::transport::Transport::Ssh{ endpoint, auth }) = integration.protocol() else { panic!("must have parsed integration to git") };
    assert_eq!(
        endpoint,
        &gen::code_remote("git@github.com:fossas/broker.git")
    );

    let api::ssh::Auth::KeyFile(file) = auth else { panic!("must have parsed ssh key file auth") };
    assert_eq!(file, &gen::path_buf("/home/me/.ssh/id_rsa"));
}

#[tokio::test]
async fn test_integration_git_ssh_key() {
    let (_, conf) = load_config!(
        "testdata/config/basic-ssh-key.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let Some(integration) = conf.integrations().as_ref().iter().next() else { panic!("must have parsed at least one integration") };
    assert_eq!(integration.poll_interval(), gen::code_poll_interval("1h"));

    let remote::Protocol::Git(remote::git::transport::Transport::Ssh{ endpoint, auth }) = integration.protocol() else { panic!("must have parsed integration") };
    assert_eq!(
        endpoint,
        &gen::code_remote("git@github.com:fossas/broker.git")
    );

    let api::ssh::Auth::KeyValue(key) = auth else { panic!("must have parsed auth value") };
    assert_eq!(key, &gen::secret("efgh5678"));
}

#[tokio::test]
async fn test_integration_git_http_basic() {
    let (_, conf) = load_config!(
        "testdata/config/basic-http-basic.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let Some(integration) = conf.integrations().as_ref().iter().next() else { panic!("must have parsed at least one integration") };
    assert_eq!(integration.poll_interval(), gen::code_poll_interval("1h"));

    let remote::Protocol::Git(remote::git::transport::Transport::Http{ endpoint, auth }) = integration.protocol() else { panic!("must have parsed integration") };
    assert_eq!(
        endpoint,
        &gen::code_remote("https://github.com/fossas/broker.git")
    );

    let Some(api::http::Auth::Basic { username, password }) = auth else { panic!("must have parsed auth value") };
    assert_eq!(username, &String::from("jssblck"));
    assert_eq!(password, &gen::secret("efgh5678"));
}

#[tokio::test]
async fn test_integration_git_http_basic_malformed_auth() {
    let (config_file_path, err) = load_config_err!(
        "testdata/config/basic-http-basic-malformed-auth.yml",
        "testdata/database/empty.sqlite"
    )
    .await;
    assert_error_stack_snapshot!(&config_file_path, err);
}

#[tokio::test]
async fn test_integration_git_http_basic_malformed_debugging() {
    let (config_file_path, err) = load_config_err!(
        "testdata/config/basic-http-basic-malformed-debugging.yml",
        "testdata/database/empty.sqlite"
    )
    .await;
    assert_error_stack_snapshot!(&config_file_path, err);
}

#[tokio::test]
async fn test_integration_git_http_header() {
    let (_, conf) = load_config!(
        "testdata/config/basic-http-header.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let Some(integration) = conf.integrations().as_ref().iter().next() else { panic!("must have parsed at least one integration") };
    assert_eq!(integration.poll_interval(), gen::code_poll_interval("1h"));
}

#[tokio::test]
async fn test_integration_git_http_no_auth() {
    let (_, conf) = load_config!(
        "testdata/config/basic-http-no-auth.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let Some(integration) = conf.integrations().as_ref().iter().next() else { panic!("must have parsed at least one integration") };
    assert_eq!(integration.poll_interval(), gen::code_poll_interval("1h"));

    let remote::Protocol::Git(remote::git::transport::Transport::Http{ endpoint, auth }) = integration.protocol() else { panic!("must have parsed integration") };
    assert_eq!(
        endpoint,
        &gen::code_remote("https://github.com/fossas/broker.git")
    );

    let None = auth else { panic!("must have parsed no auth value") };
}

#[tokio::test]
async fn test_integration_short_poll_interval() {
    let (config_file_path, err) = load_config_err!(
        "testdata/config/short-poll-interval.yml",
        "testdata/database/empty.sqlite"
    )
    .await;
    assert_error_stack_snapshot!(&config_file_path, err);
}
