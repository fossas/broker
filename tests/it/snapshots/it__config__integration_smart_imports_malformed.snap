---
source: tests/it/config.rs
expression: err
info: testdata/config/smart_imports_malformed.yml
---
load config file
├╴at {source location}
│
├─▶ parse config file in v1 format
│   ╰╴at {source location}
│
├─▶ validate parsed config file values
│   ╰╴at {source location}
│
╰─▶ validate import branches and watched branches
    ├╴at {source location}
    ├╴help: import branches must be 'true' if watched branches are provided
    ╰╴context: import branches: 'false'
